snapcraft/_schema.py

```diff
@@ -78,48 +63,4 @@ def validate(self):
             jsonschema.validate(
                 self._snapcraft, self._schema, format_checker=format_check)
         except jsonschema.ValidationError as e:
-            _handle_validation_error(e)
-
-
-def _handle_validation_error(error):
-    """Take a jsonschema.ValidationError and raise a SnapcraftSchemaError.
-
-    The validation errors coming from jsonschema are a nightmare. This function
-    tries to make them a bit more understandable.
-    """
-
-    messages = [error.message]
-
-    # error.validator_value may contain a custom validation error message. If
-    # so, use it instead of the garbage message jsonschema gives us.
-    with contextlib.suppress(TypeError, KeyError):
-        messages = [error.validator_value['validation-failure'].format(error)]
-
-    path = []
-    while error.absolute_path:
-        element = error.absolute_path.popleft()
-        # assume numbers are indices and use 'xxx[123]' notation.
-        if isinstance(element, int):
-            path[-1] = '{}[{}]'.format(path[-1], element)
-        else:
-            path.append(str(element))
-    if path:
-        messages.insert(0, "The '{}' property does not match the "
-                           "required schema:".format('/'.join(path)))
-    cause = error.cause or _determine_cause(error)
-    if cause:
-        messages.append('({})'.format(cause))
-
-    raise SnapcraftSchemaError(' '.join(messages))
-
-
-def _determine_cause(error):
-    """Attempt to determine a cause from validation error.
-
-    Returns:
-        A string representing the cause of the error (it may be empty if no
-        cause can be determined).
-    """
-
-    return _VALIDATION_ERROR_CAUSES.get(error.validator, '').format(
-        validator_value=error.validator_value)
+            raise errors.SnapcraftSchemaError.from_validation_error(e)
```

Comment:
Ah, thanks. I'll wait to see if the integration tests pass (I changed some errors) first, then I'll push.